src/jdk.jdeps/share/classes/com/sun/tools/jdeps/ClassFileReader.java

```diff
@@ -358,76 +321,23 @@ protected ClassModel readClassFile(JarFile jarfile, JarEntry e) throws IOExcepti
-            final Iterator<ClassModel> iter = new JarFileIterator(this, jarfile);
-            return () -> iter;
-        }
-    }
-
-    class JarFileIterator implements Iterator<ClassModel> {
-        protected final JarFileReader reader;
-        protected Iterator<JarEntry> entries;
-        protected JarFile jf;
-        protected JarEntry nextEntry;
-        protected ClassModel cf;
-        JarFileIterator(JarFileReader reader) {
-            this(reader, null);
-        }
-        JarFileIterator(JarFileReader reader, JarFile jarfile) {
-            this.reader = reader;
-            setJarFile(jarfile);
-        }
-
-        void setJarFile(JarFile jarfile) {
-            if (jarfile == null) return;
-
-            this.jf = jarfile;
-            this.entries = jarfile.versionedStream().iterator();
-            this.nextEntry = nextEntry();
-        }
-
-        public boolean hasNext() {
-            if (nextEntry != null && cf != null) {
-                return true;
-            }
-            while (nextEntry != null) {
-                try {
-                    cf = reader.readClassFile(jf, nextEntry);
-                    return true;
-                } catch (ClassFileError | IOException ex) {
-                    skippedEntries.add(String.format("%s: %s (%s)",
-                                                     ex.getMessage(),
-                                                     nextEntry.getName(),
-                                                     jf.getName()));
-                }
-                nextEntry = nextEntry();
-            }
-            return false;
-        }
-
-        public ClassModel next() {
-            if (!hasNext()) {
-                throw new NoSuchElementException();
-            }
-            ClassModel classFile = cf;
-            cf = null;
-            nextEntry = nextEntry();
-            return classFile;
-        }
-
-        protected JarEntry nextEntry() {
-            while (entries.hasNext()) {
-                JarEntry e = entries.next();
-                String name = e.getName();
-                if (name.endsWith(".class")) {
-                    return e;
-                }
+        @Override
+        public void processClassFiles(Consumer<ClassModel> handler) throws IOException {
+            try {
+                jarfile.versionedStream()
+                        .filter(JarFileReader::isJarEntryClass)
+                        .forEach(e -> {
+                            try {
+                                handler.accept(readClassFile(jarfile, e));
+                            } catch (ClassFileError ex) {
+                                skipEntry(ex, e.getName() + " (" + jarfile.getName() + ")");
+                            } catch (IOException ex) {
+                                throw new UncheckedIOException(ex);
+                            }
+                        });
+            } catch (UncheckedIOException ex) {
+                throw ex.getCause();
```

Comment:
IOException used to skip entry with message and continue, the new behavior would change. I am not certain this is behavior compatible.
I doubt IOException would be recoverable on different entry, but it might in rare cases?

Reply:
Yep, I think skipping for IOE is correct. If IOE is serious enough that no element is accessible, it should usually be thrown earlier.